near/omni-prover/evm-prover/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::{env, ext_contract, near_bindgen, AccountId, Gas, PanicOnDefault, Promise};
+use omni_types::evm::events::parse_evm_event;
+use omni_types::evm::header::BlockHeader;
+use omni_types::evm::receipt::{LogEntry, Receipt};
+use omni_types::evm::utils::keccak256;
+use omni_types::prover_args::EvmVerifyProofArgs;
+use omni_types::prover_result::{ProofKind, ProverResult};
+use omni_types::ChainKind;
+use rlp::Rlp;
+
+const VERIFY_PROOF_CALLBACK_GAS: Gas = Gas::from_tgas(20);
+const BLOCK_HASH_SAFE_GAS: Gas = Gas::from_tgas(10);
+
+type H256 = [u8; 32];
+
+/// Defines an interface to call EthClient contract to get the safe block hash for a given block
+/// number. It returns Some(hash) if the block hash is present in the safe canonical chain, or
+/// None if the block number is not part of the canonical chain yet.
+#[ext_contract(evm_client)]
+pub trait EvmClient {
+    #[result_serializer(borsh)]
+    fn block_hash_safe(&self, #[serializer(borsh)] index: u64) -> Option<H256>;
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct EvmProver {
```

Comment:
Do we want to add roles? The ability to pause the contract? The ability to change the light_client?

Or are we planning to simply deploy a new contract and change the prover in the higher-level contract if necessary?

Reply:
we planning to simply deploy a new contract and change the prover in the higher-level contract